### ISSUE TITLE
script_phase_constants contains bashisms but tries to use /bin/sh

### ISSUE THREAD

**Issue by yangm97** [CocoaPods/CocoaPods#10838]
https://github.com/CocoaPods/CocoaPods/blob/42cce8842efa64c15f3510cf97b4d9e961ee85b7/lib/cocoapods/generator/script_phase_constants.rb#L5

As of now bash is used as the default shell in macOS, but it is not very strict about POSIX standards. It may be the case that, in the future, given that Apple is moving away from bash, it could be replaced by something stricter, [dash maybe?](https://scriptingosx.com/2020/06/about-bash-zsh-sh-and-dash-in-macos-catalina-and-beyond/).

You can confirm this by switching from bash to dash as the sh interpreter, I did that in order to catch bashisms within my own scripts:

```
sudo ln -sf /bin/dash /var/select/sh 
```

In order to prevent issues in the future we should either get rid of the bashisms or change those script files to call bash explicitly but hardcode its path, like this:

```
#/usr/bin/env bash
```

**Comment by dnkoutso** [comment]
I think that's fine, want to make a PR for it? we'd need to update integration specs but thats easy.

**Comment by yangm97** [comment]
Sure, but how exactly do you see this fixed?

a. Make scripts POSIX compliant: requires some work and continuous attention (i.e. adding shellcheck to the CI tests in order to prevent bashisms from popping up again)

b. Pick bash from env: has a chance of breaking people's workflow __today__, since it's not guaranteed to pick up the system bash (the user could've installed a more up to date bash from brew or whatever)

c. Hardcode /bin/bash: has a chance of breaking very very very far in the future, if apple does in fact nuke bash from macOS 

d. Hardcode /bin/zsh: since it has became the default shell it shouldn't be going anywhere

e. ?

**Comment by dnkoutso** [comment]
isnt c) the current option?

I wouldnt prefer b) as it will cause a huge rout to folks. we can do b) only if we can provide a backup.

Wouldnt d) also break some folks?

**Comment by yangm97** [comment]
> isnt c) the current option?

Not exactly, we're implicitly using bash in sh compatibility mode, which isn't very strict as you can see but is different from calling bash directly. So even c. has some, albeit low, risk involved.

> I wouldnt prefer b) as it will cause a huge rout to folks. we can do b) only if we can provide a backup.

wdym by providing a backup? 

> Wouldnt d) also break some folks?

There is indeed a risk of someone getting caught in some corner case. It could also be the case that, just like a. we could need to modify the scripts to make them zsh compatible, further increasing the chances of breaking.

---

The very possibility of b. breaking people’s workflows kind of stems from the fact bundled bash is ancient and heavily patched by Apple. I really wouldn't be surprised if it got axed the same way bundled ruby, python and perl are supposed to. 

So while c. supposedly offers the lowest risk and would fix my personal issue _right now_ I think it's fair to look for a long term solution, if possible. 

I'm also trying to stand on the theoretical ground as much as possible because I don't know the inner workings of cocoapods nor the nature of these scripts.
